src/csharp/Grpc.Core/Internal/AsyncCall.cs

```diff
@@ -116,7 +116,8 @@ public TResponse UnaryCall(TRequest msg)
                             {
                                 using (profiler.NewScope("AsyncCall.UnaryCall.HandleBatch"))
                                 {
-                                    HandleUnaryResponse(success, ctx.GetReceivedStatusOnClient(), ctx.GetReceivedMessageReader(), ctx.GetReceivedInitialMetadata());
+                                    HandleUnaryResponse(success, ctx.TryGetReceivedStatusOnClient(),
```

Comment:
Changing the way the data gets read from context here would make the behavior for unary calls and for streaming calls inconsistent and I don't think we want that. I think instead, we should make a small modification to the original GetReceivedStatusOnClient() method and revert the changes here.

Reply:
The change is to make sure no exceptions are thrown in `UnaryCall`  to avoid a hang (by wrapping _all_ methods that are being called before calling `HandleUnaryResponse`) and make sure that the consequences of any exception are cleaned up.

In the original `HandleUnaryResponse` there was already a `TryDeserialize` - I am just applying the same pattern to the other methods `TryGetReceivedStatusOnClient` and `TryGetReceivedInitialMetadata`.